addons/pos_viva_com/static/src/app/hooks/use_viva_app.js

```diff
@@ -0,0 +1,143 @@
+            const message = urlParams.get("message");
+            line?.delete();
+            dialog.add(AlertDialog, {
+                title: _t("Viva Wallet Payment Error"),
+                body: `Please note that your order has not been finalized, try again or choose another payment method. (${message})`,
+            });
+        }
+
+        // Remove the URL parameters to avoid processing them again
+        const newUrl = window.location.href.split("?")[0];
+        window.history.replaceState({}, document.title, newUrl);
+    };
+
+    /**
+     * This function is used to open the Viva Wallet app
+     * and start the payment process. It will open the app with the
+     * necessary parameters to process the payment.
+     *
+     * @param {Object} paymentMethod
+     * @param {Boolean} isRefund
+     */
+    const start = async (paymentMethod, isRefund = false) => {
+        const line = order.addPaymentline(paymentMethod);
+        try {
+            line.viva_com_session_id = `${order.uuid}-${uuidv4()}`;
+            const result = await pos.syncAllOrders({ orders: [order] });
+            await pos.data.synchronizeLocalDataInIndexedDB();
+            if (!result) {
+                throw new Error(
+                    "Impossible to initiate Vivawallet payment without syncing the order"
+                );
+            }
+
+            let url =
+                "vivapayclient://pay/v1" +
+                "?appId=com.example.odoo" + // We don't have a real appId, so we use a dummy one
+                `&amount=${roundPrecision(Math.abs(line.amount * 100))}` +
+                "&show_receipt=true" +
+                "&show_transaction_result=true" +
+                "&show_rating=true" +
+                `&callback=${pos.session._base_url}/pos/ui/${pos.config.id}/payment/${order.uuid}`;
+
+            if (isRefund) {
+                const refundedOrderPay = line.pos_order_id.refunded_order_id.payment_ids;
+                const vivaLine = refundedOrderPay.find(
+                    (p) => p.payment_method_id.use_payment_terminal === "viva_com"
+                );
+
+                if (vivaLine) {
+                    url += `&referenceNumber=${vivaLine.transaction_id}`;
+                }
+
+                url += `&action=cancel`;
+            } else {
+                url += `&action=sale`;
+                url += `&clientTransactionId=${line.viva_com_session_id}`;
+                url += `&ISV_currencyCode=${pos.currency.iso_numeric.toString()}`;
+                url += `&ISV_merchantId=${line.viva_com_session_id}/${pos.session.id}`;
+                url += `&tipAmount=0`;
+                url += `&paymentMethod=CardPresent`;
+
+                if (order.partner) {
+                    url += `&ISV_customerTrns=${order.partner.name}-${order.partner.email}`;
+                }
+            }
+
+            line.setPaymentStatus("waitingCard");
+            window.open(url, "_self");
```

Comment:
Will this reload the whole pos app? If that's the case, for db's with slow pos-loading, then this feature won't be usable.

Isn't it possible to open the vivapaymentclient url in a different tab and when the payment is done, notify the tab that contains the pos app? We already have this trick of knowing to close an open pos if the same pos is opened in a different tab.

Reply:
isn't a tab is an android URL which will open an app